cidc_api/models/templates/csms_api.py

```diff
@@ -500,3 +549,497 @@ def insert_manifest_from_json(
+                )
+                if db_sample is not None
+                else f"<no sample found>"
+            )
+            raise Exception(
+                f"Change in critical field for: {formatted} to CSMS {(trial_id, manifest_id, cimac_id)}"
+            )
+
+    # add here for ordering and then just append; remove later if not needed
+    ret0[Participant] = []
+    ret0[Sample] = []
+    for cimac_id, csms_sample in csms_sample_map.items():
+        diff = _calc_difference(cidc_sample_map[cimac_id], csms_sample)
+        if diff:
+            ret1["samples"].append(diff)
+
+            # pull out just the changes for ease
+            changes = set(diff.keys())
+            for i in ["cimac_id", "shipment_manifest_id", "trial_id"]:
+                changes.remove(i)
+
+            # Participant-level fields
+            if any(
+                i in changes
+                for i in ["cohort_name", "participant_id", "trial_participant_id"]
+            ):
+                # since insert_record_batch is a merge, we can just generate a new object
+                new_partic = Participant(
+                    trial_participant_id=csms_sample["participant_id"],
+                    **_get_all_values(target=Participant, old=csms_sample),
+                )
+                ret0[Participant].append(new_partic)
+            for i in ["cohort_name", "participant_id", "trial_participant_id"]:
+                if i in changes:
+                    changes.remove(i)
+
+            # if there's only cohort_name or trial_participant_id, there's nothing left
+            if len(changes):
+                # since insert_record_batch is a merge, we can just generate a new object
+                new_sample = Sample(**_get_all_values(target=Sample, old=csms_sample))
+                ret0[Sample].append(new_sample)
+
+    # ----- Look for differences in the Upload -----
+    cidc_upload = (
+        session.query(Upload).filter(Upload.shipment_manifest_id == manifest_id).first()
+    )
+    new_upload = Upload(
+        trial_id=trial_id,
+        status=UploadJobStatus.MERGE_COMPLETED.value,
+        multifile=False,
+        assay_creator="DFCI",
+        upload_type=_get_upload_type(csms_samples),
+        shipment_manifest_id=manifest_id,
+        uploader_email=user.email,
+    )
+    diff = _calc_difference(
+        {} if cidc_upload is None else cidc_upload.to_dict(),
+        new_upload.to_dict(),
+        ignore=["id", "token", "uploader_email"],
+    )
+    if diff:
+        # since insert_record_batch is a merge, we can just generate a new object
+        ret0[Upload] = [new_upload]
+        ret1["upload"].append(diff)
+
+    # ----- Finish up and return -----
+    # drop unneeded keys
+    if len(ret0[Participant]) == 0:
+        ret0.pop(Participant)
+    if len(ret0[Sample]) == 0:
+        ret0.pop(Sample)
+
+    # convert away from defaultdict
+    return ret0, dict(ret1)
+
+
+@with_default_session
+def update_with_changes(
```

Comment:
I guess my preference would be to leave this out of scope for this PR. Could you stash this for a future sprint?

Reply:
as this is not currently connected to anything and have full testing written, I think it's worth adding it now

Reply:
otherwise I can pull this and its tests onto a separate branch to merge later